crates/adapters/src/transport/secret_resolver.rs

```diff
@@ -0,0 +1,260 @@
+use regex::Regex;
+use std::fmt;
+use std::fmt::{Debug, Display, Formatter};
+use std::fs;
+
+use log::{debug, warn};
+use pipeline_types::secret_ref::SimpleStringOrSecretRef;
+
+/// Enumeration which holds a simple string or a resolved secret's string.
+///
+/// This enumeration exists such that a resolved secret's string is not shown
+/// upon formatting (irrespective if `Display` or `Debug`). This is preferable
+/// over using a simple `String`, which would show its content upon formatting.
+///
+/// In addition, a warning is emitted when a secret is attempted to be
+/// formatted.
+#[derive(Clone, PartialEq, Eq)]
+pub enum SimpleOrSecretString {
+    Simple(String),
+    Secret(String),
+}
+
+impl Display for SimpleOrSecretString {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            SimpleOrSecretString::Simple(simple_string) => {
+                write!(f, "Simple({})", simple_string)
+            }
+            SimpleOrSecretString::Secret(_) => {
+                warn!(
+                    "Resolved secret is requested to be formatted. \
+                       It is not allowed to format the internal secret string, \
+                       and as such it is not shown."
+                );
+                write!(f, "Secret(***)")
+            }
+        }
+    }
+}
+
+impl Debug for SimpleOrSecretString {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        // Use Display formatting
+        write!(f, "{}", self)
+    }
+}
+
+impl SimpleOrSecretString {
+    /// A secret reference can only contain a-z and 0-9. There can be connecting
+    /// hyphens (-), though not at the start, not at the end, and not two in
+    /// succession. It cannot be empty.
+    ///
+    /// Valid examples:
+    /// "example", "example-1", "example-of-this"
+    ///
+    /// Invalid examples:
+    /// "", "example--1", "example_1", "-", "Example#1", "example-"
+    fn is_valid_secret_ref_format(secret_ref: &str) -> bool {
+        let re = Regex::new(r"^[a-z0-9]+(-[a-z0-9]+)*$").unwrap();
+        re.is_match(secret_ref)
+    }
+
+    pub fn new_with_default_volume(value: SimpleStringOrSecretRef) -> Self {
+        Self::new("/etc/secret-volume".to_string(), value)
```

Comment:
if the new `secret_volumn_path` is of type `AsRef<str>` it can probably work for both &str and String types

Reply:
Changed the type of the `new()` argument to `&Path` instead of a string. I kept the `String` for the `SecretRef` class with the motivation to keep the semantic there is some transfer of ownership, rather than turning it into e.g., `&dyn AsRef<str>`. Another approach could be `impl Into<String>`. Is that motivation reasonable? I can also change it, no problem.